src/engine/Operation.cpp

```diff
@@ -134,47 +138,112 @@ shared_ptr<const ResultTable> Operation::getResult(bool isRoot,
+        result.enforceLimitOffset(_limit);
+      }
+      return result;
+    };
+
+    auto cacheSetup = [this, &computeLambda, &cache, &cacheKey]() {
+      auto result = computeLambda();
+      if (!result.isDataEvaluated()) {
+        result.setOnSizeChanged([&cache, cacheKey](bool isShrinkable) {
+          // TODO<RobinTF> find out how to handle pinned entries properly.
+          auto sizeChange = cache.recomputeSize(cacheKey, !isShrinkable);
+          if (sizeChange ==
+              ad_utility::ResizeResult::EXCEEDS_SINGLE_ENTRY_SIZE) {
+            return isShrinkable;
+          }
+          return false;
+        });
+        result.setOnGeneratorFinished(
+            [&cache, cacheKey](bool isComplete) mutable {
+              if (isComplete) {
+                cache.transformValue(cacheKey, [](const CacheValue& oldValue) {
+                  return CacheValue{oldValue.resultTable()->aggregateTable(),
+                                    oldValue.runtimeInfo()};
+                });
+              }
+            });
+        result.setOnNextChunkComputed([runtimeInfo = getRuntimeInfoPointer()](
+                                          std::chrono::milliseconds duration) {
+          runtimeInfo->totalTime_ += duration;
+        });
       }
       return CacheValue{std::move(result), runtimeInfo()};
     };
 
-    auto result = (pinResult) ? cache.computeOncePinned(cacheKey, computeLambda,
-                                                        onlyReadFromCache)
-                              : cache.computeOnce(cacheKey, computeLambda,
-                                                  onlyReadFromCache);
+    bool onlyReadFromCache = computationMode == ComputationMode::CACHE_ONLY;
+
+    auto result =
+        pinResult
+            ? cache.computeOncePinned(cacheKey, cacheSetup, onlyReadFromCache)
+            : cache.computeOnce(cacheKey, cacheSetup, onlyReadFromCache);
 
     if (result._resultPointer == nullptr) {
       AD_CORRECTNESS_CHECK(onlyReadFromCache);
       return nullptr;
     }
 
-    updateRuntimeInformationOnSuccess(result, timer.msecs());
-    auto resultNumRows = result._resultPointer->resultTable()->size();
-    auto resultNumCols = result._resultPointer->resultTable()->width();
-    LOG(DEBUG) << "Computed result of size " << resultNumRows << " x "
-               << resultNumCols << std::endl;
-    return result._resultPointer->resultTable();
+    if (result._resultPointer->resultTable()->isDataEvaluated()) {
+      updateRuntimeInformationOnSuccess(result, timer.msecs());
+    }
+
+    if (result._resultPointer->resultTable()->isDataEvaluated()) {
+      auto resultNumRows =
+          result._resultPointer->resultTable()->idTable().size();
+      auto resultNumCols =
+          result._resultPointer->resultTable()->idTable().numColumns();
+      LOG(DEBUG) << "Computed result of size " << resultNumRows << " x "
+                 << resultNumCols << std::endl;
+    }
```

Comment:
Does this debug message provide any real benefit to make it worth somehow incorporating it into lazily evaluated operations?